### PR TITLE
[On hold] New formulas for derivatives of neg_binomial_lpmf after alpha cutoff

### PR DESCRIPTION
**Pull request by martinmodrak**
## Summary

*Currently on-hold. If PR #1497 gets merged, `neg_binomial_lpmf` should probably be rewritten using the same approach*

The formulas for derivatives of `neg_binomial_lpmf` after delegating to poisson (for `alpha > 1e10`) are replaced with Taylor expansions of the exact derivatives for `alpha -> Inf` obtained in Mathematica via:

```
     nb[n_,alpha_,beta_]:= LogGamma[n + alpha] - LogGamma[n + 1] - 
        LogGamma[alpha ] + alpha * Log[beta/ (1 + beta)] - n * Log[1 + beta];
      nbdalpha[n_,alpha_,beta_]= D[nb[n, alpha, beta],alpha];
      nbdbeta[n_,alpha_,beta_]= D[nb[n, alpha, beta],beta];
      Series[nbdalpha[n, alpha, beta],{alpha, Infinity, 1}]
      Series[nbdbeta[n, alpha, beta],{alpha, Infinity, 0}]
```

(the lowest order of the series that let's me pass the tests was chosen)

## Tests

Tests are adapted from #1497 where new tests for `neg_binomial_2_lpmf` were written. Namely

- Created a set of test values and derivatives in Mathematica (Wolfram Cloud) and test against those
- Testing the derivatives against complex step derivatives. This is important for large `alpha`, as Mathematica refuses to compute exact derivatives in this case.
- Over a grid of `beta` and `n` values test whether the change in the derivatives of `neg_binomial_lpmf` at `alpha_cutoff +/- 1e-8` is negligible.
- Test whether there is no drastic change in `neg_binomial_lpmf<propto=true, int, var, double>` between `alpha` just below and just above the cutoff.

## Side Effects

Not any.

## Checklist

- [x] Math issue #1578 

- [x] Copyright holder: Institute of Microbiology of the Czech Academy of Sciences

    By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

**Comment by martinmodrak** [comment]
@bob-carpenter May I ask for review on this? The tests here are similar to those written for #1497 (although the problem addressed is different).

File: stan/math/prim/prob/neg_binomial_lpmf.hpp [hunk 1]
```diff
@@ -15,6 +15,12 @@
 namespace stan {
 namespace math {
 
+namespace internal {
+// Exposing to let me use in tests
```
Comment (bob-carpenter) [inline stan/math/prim/prob/neg_binomial_lpmf.hpp]: Comments shouldn't be first person.  I'd suggest just eliminating line 19.

Line 20 should say what the value is for---in this case, the threshold beyond which a Poisson approximation is used.  It's also OK to say why it's 1e10, but I couldn't follow the logic of why it's set to 1e10.

File: stan/math/prim/prob/neg_binomial_lpmf.hpp [hunk 2]
```diff
@@ -102,19 +108,30 @@ return_type_t<T_shape, T_inv_scale> neg_binomial_lpmf(const T_n& n,
   }
 
   for (size_t i = 0; i < max_size_seq_view; i++) {
-    if (alpha_vec[i] > 1e10) {  // reduces numerically to Poisson
+    if (alpha_vec[i] > internal::neg_binomial_alpha_cutoff) {
+      // reduces numerically to Poisson
+      // The derivatives are obtained via Taylor series at alpha -> Inf
+      // via Mathematica as:
+      // nb[n_,alpha_,beta_]:= LogGamma[n + alpha] - LogGamma[n + 1] -
```
Comment (bob-carpenter) [inline stan/math/prim/prob/neg_binomial_lpmf.hpp]: I don't think we should assume the readers of comments understand mathematica.  So this should be put in Stan notation or just eliminated.  I couldn't understand what `LogGamma` is or `Log` or `D`.  If these comments stay, the functions used need to be explained.

File: stan/math/prim/prob/neg_binomial_lpmf.hpp (hunk 1 above)
Comment (syclik) [reply stan/math/prim/prob/neg_binomial_lpmf.hpp]: @bob-carpenter, would it be ok if L20 said something like:
``` // Set to 1e10 based on empirical evaluation. At 1e9, the approximation isn't close enough```

**Comment by bob-carpenter** [comment]
> 
> @bob-carpenter, would it be ok if L20 said something like:
> // Set to 1e10 based on empirical evaluation. At 1e9, the approximation isn't close enough

Yes, that's clear.

**Comment by martinmodrak** [comment]
Thx @syclik and @bob-carpenter for the notes. I should've marked this PR as "on hold" because if #1497 turns out to be on the right track, I would probably rewrite `neg_binomial_lpmf` in the same spirit (eliminating the Poisson branch completely). So sorry for wasting your time here.

**Comment by syclik** [comment]
@martinmodrak, do you want to close this PR? #1497 was merged and from reading through the PR, it seems like you are favoring a different implementation.

**Comment by martinmodrak** [comment]
Thanks for reminidng me, yes this should be closed.

**Pull request closed without merging**